nemo_skills/inference/generate.py

```diff
@@ -257,48 +259,92 @@ def async_loop(cfg, data, llm, prompt, extra_stop_phrases, extra_generate_params
+            idx = request_queue.popleft()
+            in_progress[idx] = None  # Placeholder for reservation
 
+        # **Step 2: Submit first max_concurrent_requests requests as a batch**
+        batch_indices = list(in_progress.keys())  # Get reserved indices
+        batch_prompts = [prompt.fill(data[idx]) for idx in batch_indices]  # Prepare prompts
+
+        batch_ids = llm.generate_async(  # Submit all at once
+            prompts=batch_prompts,  
+            stop_phrases=combine_stop_phrases(prompt.stop_phrases, extra_stop_phrases),
+            **asdict(cfg.inference),
+            **extra_generate_params,
+        )
+
+        # Store the request IDs returned by LLM
+        for i, idx in enumerate(batch_indices):
+            in_progress[idx] = batch_ids[i]
+
+        # **Step 3: Monitor and refill requests dynamically**
+        while in_progress or request_queue:  # Continue until all tasks are complete
+            remaining_ids = {idx: gen_id for idx, gen_id in in_progress.items()}
+            generations = llm.get_generations(list(remaining_ids.values()))
+            
+            completed_tasks = []
+            for (idx, gen_id), gen_dict in zip(remaining_ids.items(), generations):
+                if gen_dict['generation'] is not None:
+                    # Mark task as completed
+                    completed_tasks.append(idx)
+                    del in_progress[idx]
+                    ### we can remove the gen_id from the gen_id_to_params to save memory when we are done with it
+                    if gen_id in llm.gen_id_to_params:
+                        del llm.gen_id_to_params[gen_id]
+                        
+                    # Prepare the result for writing
+                    gen_dict[cfg.generation_key] = gen_dict.pop("generation")
+                    for key in gen_dict:
+                        data[idx].pop(key, None)
+                    gen_dict.update(data[idx])
+                    
+                    # Insert the async position to preserve the original order
+                    gen_dict[cfg.async_position_key] = original_positions[idx]
+                    
+                    # Write the result immediately to minimize memory usage
+                    fout.write(json.dumps(gen_dict) + "\n")  
+
+                    # Update progress bar
+                    pbar.update(1)
+
+
+
+            
+            # **Step 4: Refill requests to maintain exactly N=max_concurrent_requests concurrent tasks**
+            # num_to_submit = cfg.max_concurrent_requests - len(in_progress)
+            # batch_indices = [request_queue.popleft() for _ in range(min(num_to_submit, len(request_queue)))]
+            num_to_submit = min(cfg.max_concurrent_requests - len(in_progress), len(request_queue))
+            batch_indices = [request_queue.popleft() for _ in range(num_to_submit)]
+
+            batch_prompts = [prompt.fill(data[idx]) for idx in batch_indices]
```

Comment:
Right now the prompt_fill and generation part are repeated before the while loop and at the end of the while loop. I would suggest removing this duplication by moving this piece of code to start of the while loop.

Reply:
I meant the piece of code is repetitive.